### PR TITLE
Store operations in shared storage.

### PR DESCRIPTION
**Pull request by gterzian**
Fix https://github.com/massalabs/massa/issues/2582

### CODE REVIEW

**Comment by gterzian** [comment]
@massalabs/core-team Update on this: adding the network integration, should be ready for a review tomorrow...

**Comment by gterzian** [comment]
This one is now ready for review. @massalabs/core-team

**Comment by gterzian** [comment]
As discussed, the last commit removes the inner lock, and adds some optimizations to avoid cloning when we can...

File: massa-protocol-exports/src/protocol_controller.rs
Comment (adrien-zinger) [inline massa-protocol-exports/src/protocol_controller.rs:50]: You can use `OperationIds`

File: massa-protocol-exports/src/tests/mock_network_controller.rs
Comment (AurelienFT) [inline massa-protocol-exports/src/tests/mock_network_controller.rs:99]: Why we don't store real serialize operations?

File: massa-network-worker/src/node_worker.rs
Comment (gterzian) [reply massa-network-worker/src/node_worker.rs:420]: I'm running into the below, think we would have to implement `From` for some of the prehashed structures, worth looking into in a follow-up if you think it's worth it...

```
error[E0277]: the trait bound `HashSet<OperationId, BuildHasherDefault<PreHashedMap<OperationId>>>: From<Chunk<'_, std::collections::hash_set::IntoIter<OperationId>>>` is not satisfied
   --> massa-network-worker/src/node_worker.rs:422:103
    |
422 | ...                   if self.try_send_to_node(&writer_command_tx, ToSend::Operations(chunk.into())).is_err() {
    |                                                                                             ^^^^ the trait `From<Chunk<'_, std::collections::hash_set::IntoIter<OperationId>>>` is not implemented for `HashSet<OperationId, BuildHasherDefault<PreHashedMap<OperationId>>>`
    |
    = help: the following implementations were found:
              <HashSet<T> as From<[T; N]>>
    = note: required because of the requirements on the impl of `Into<HashSet<OperationId, BuildHasherDefault<PreHashedMap<OperationId>>>>` for `Chunk<'_, std::collections::hash_set::IntoIter<OperationId>>`
```

File: massa-pool/src/operation_pool.rs
Comment (gterzian) [reply massa-pool/src/operation_pool.rs:41]: Good one, will do.

File: massa-pool/src/operation_pool.rs
Comment (gterzian) [reply massa-pool/src/operation_pool.rs:136]: It's what is done already I think.

File: massa-protocol-exports/src/tests/mock_network_controller.rs
Comment (gterzian) [reply massa-protocol-exports/src/tests/mock_network_controller.rs:99]: Because the serialized form is unrelated to the protocol tests. You can see it as a kind of mock.

File: massa-protocol-worker/src/worker_operations_impl.rs
Comment (gterzian) [reply massa-protocol-worker/src/worker_operations_impl.rs:204]: It's debatable, and I think it's better to ask pool, because then we can assert that the operations are in fact present in storage, otherwise we cannot(in protocol), see the difference between "not there but should be there(error)" and "not there because it's not supposed to be there".

File: massa-protocol-worker/src/worker_operations_impl.rs
Comment (gterzian) [reply massa-protocol-worker/src/worker_operations_impl.rs:142]: They are still separate concepts, and the option could be removed if we do https://github.com/massalabs/massa/issues/2622

File: massa-protocol-exports/src/protocol_controller.rs
Comment (gterzian) [reply massa-protocol-exports/src/protocol_controller.rs:50]: Ok thanks

File: massa-network-worker/src/network_event.rs
Comment (gterzian) [reply massa-network-worker/src/network_event.rs:236]: That is done by protocol, after performing some additional checks on the operation, so network does not store it yet at this point.

File: massa-network-worker/src/messages.rs
Comment (gterzian) [reply massa-network-worker/src/messages.rs:73]: I think it's kind of the same as wrapping `SerializedForm` inside an option as now, actually I think it's clearer because having a `None` variant could mean "does not have a serialized form", where in fact it does, but we just don't use it.

File: massa-network-worker/src/node_worker.rs
Comment (gterzian) [reply massa-network-worker/src/node_worker.rs:420]: Filed https://github.com/massalabs/massa/issues/2623

File: massa-pool/src/operation_pool.rs
Comment (AurelienFT) [reply massa-pool/src/operation_pool.rs:136]: Ah yes it's done in the inner function.

**Comment by AurelienFT** [comment]
The definition of the trait of Serializer and Deserializer and usages are in #2625

File: massa-network-worker/src/messages.rs
Comment (adrien-zinger) [reply massa-network-worker/src/messages.rs:73]: Yes, indeed it's more idiomatic. I just think at loud about the format _Enum in Enum_

File: massa-network-worker/src/network_event.rs
Comment (adrien-zinger) [reply massa-network-worker/src/network_event.rs:236]: Ok!

File: massa-protocol-worker/src/worker_operations_impl.rs
Comment (adrien-zinger) [reply massa-protocol-worker/src/worker_operations_impl.rs:204]: Because we're dealing with mpsc timeouts, I thought that we would like to reduce the communication between modules. That said, I understand your point... if we store the operations inside _pool_ instead of _protocol_, would we be sure that a load failed only in the case _"not there because it's not supposed to be there"_?  

About that, a distant node never ask another node something not declared with some batch ids.

So in a nominal case, we should never fail to load operations, right? And if a malicious (:smirk:)  ask us something that we don't have, both _"not there but should be"_ and _"not supposed to be here"_ result to the same behavior?

Comment (gterzian) [reply massa-protocol-worker/src/worker_operations_impl.rs:204]: Yes, I think it could be a nice follow-up, especially if we can somehow know for sure that "it doesn't matter" from a logical perspective.

**Comment by gterzian** [comment]
I will apply the same fix as in https://github.com/massalabs/massa/pull/2629 here.

**Comment by gterzian** [comment]
@massa-core @adrien-zinger @ AurelienFT @damip @Eitu33 This one is ready for review. Sorry I cannot make a presentation today because I am having some connection problems...

**Comment by gterzian** [comment]
Rebased to fix the clippy error.

**Review by AurelienFT — APPROVED** [review]
It's ok for me.

**Comment by gterzian** [comment]
@AurelienFT Rebased to fix the clippy error. Can we get a re-approval please? I think it should be good to merge unless someone still wants to take a look @massalabs/core-team

**Review by AurelienFT — APPROVED** [review]

**Comment by damip** [comment]
Let's not merge immediately, wait for 11.1 and then we will merge it (should be today)

**Comment by damip** [comment]
bors merge

**Pull request merged**
